### PR TITLE
Allow exposure data driven classification

### PR DESCRIPTION
**Pull request by mbernasocchi**
### What does it fix?
* Ticket: adresses #3626 
* Funded by: WB GFDRR
* Description: 
This PR introduces a new classification for exposure data. This classification is data driven, meaning that classes are automatically generated based on the unique values present in the dataset.

image

image

image

image

### CODE REVIEW

File: safe/gui/tools/wizard/step_kw40_classify.py [hunk 1]
```diff
@@ -169,16 +168,33 @@ def set_widgets(self):
...
+
+        # get default classes
+        default_classes = classification['classes']
+        if classification['key'] == 'data_driven_classes':
+            default_classes = []
+            for unique_value in clean_unique_values:
+                default_class = {'key': unique_value,
+                                 # 'name': tr('Settlement'),
+                                 # 'description': tr('Settlement'),
+                                 'string_defaults': [unicode(unique_value).upper().replace('_', ' ')]}
```
Comment (ismailsunni) [inline safe/gui/tools/wizard/step_kw40_classify.py]: __[flake8]__
* [F821] undefined name 'unicode'

**Comment by mbernasocchi** [comment]
@sopac fyi

File: safe/gis/vector/prepare_vector_layer.py [hunk 2]
```diff
@@ -166,7 +166,11 @@ def _check_value_mapping(layer, exposure_key=None):
         classification = layer.keywords['classification']
 
     exposure_classification = definition(classification)
-    other = exposure_classification['classes'][-1]['key']
+
+    if exposure_classification['key'] == 'data_driven_classes':
+        other = 'other'
```
Comment (Gustry) [inline safe/gis/vector/prepare_vector_layer.py]: AFAIK, this is not necessary as you have your own data driven classes. The other group shouldn't be used, as you are mapping each one to its own class.

File: safe/gui/tools/wizard/step_kw40_classify.py [hunk 3]
```diff
@@ -263,6 +277,7 @@ def populate_classified_values(
         :param default_classes: Default classes from unit.
         :type default_classes: list
         """
+
```
Comment (Gustry) [inline safe/gui/tools/wizard/step_kw40_classify.py]: We usually don't add new line after a docstring (see the landscape analysis)

File: safe/gis/vector/prepare_vector_layer.py (hunk 2 above)
Comment (mbernasocchi) [inline safe/gis/vector/prepare_vector_layer.py]: exposure_classification['classes'][-1] will fail when classes = []

File: safe/gui/tools/wizard/step_kw40_classify.py [hunk 4]
```diff
@@ -169,16 +168,35 @@ def set_widgets(self):
...
+        # get default classes
+        default_classes = classification['classes']
+        if classification['key'] == 'data_driven_classes':
+            default_classes = []
+            for unique_value in clean_unique_values:
+                default_class = {'key': unique_value,
+                                 # 'name': tr('Settlement'),
+                                 # 'description': tr('Settlement'),
+                                 'string_defaults':
+                    [unicode(unique_value).upper().replace('_', ' ')]
```
Comment (ismailsunni) [inline safe/gui/tools/wizard/step_kw40_classify.py]: __[flake8]__
* [F821] undefined name 'unicode'

**Comment by timlinux** [comment]
Very nice @mbernasocchi - won't it make sense to just skip the wizard step as per your last screenshot above - it there any point in making the user still go into it if everything is already mapped?

File: safe/gui/tools/wizard/step_kw40_classify.py (hunk 1 above)
Comment (mbernasocchi) [inline safe/gui/tools/wizard/step_kw40_classify.py]: @ismailsunni there are plenty of other usages of unicode in that module

File: safe/gis/vector/prepare_vector_layer.py (hunk 2 above)
Comment (mbernasocchi) [inline safe/gis/vector/prepare_vector_layer.py]: I changed the logic a bit

**Comment by mbernasocchi** [comment]
@timlinux well, so you can still tweak it if needed. It almost allows on the fly creation of classifications. I would live it that way, but open for change.

File: safe/gui/tools/wizard/step_kw40_classify.py [hunk 5]
```diff
@@ -169,16 +168,34 @@ def set_widgets(self):
...
                     unique_value, QPyNullVariant):
                 # Don't classify features with NULL value
                 continue
+            clean_unique_values.append(unique_value)
+
+        # get default classes
+        default_classes = classification['classes']
+        if classification['key'] == 'data_driven_classes':
+            for unique_value in clean_unique_values:
+                name = unicode(unique_value).upper().replace('_', ' ')
```
Comment (ismailsunni) [inline safe/gui/tools/wizard/step_kw40_classify.py]: __[flake8]__
* [F821] undefined name 'unicode'

**Comment by timlinux** [comment]
@mbernasocchi please add new features to the changelog here http://changelog.qgis.org/en/inasafe/version/4.2.0/ as you add them. Thanks!

**Comment by ismailsunni** [comment]
Yeay

**Comment by ismailsunni** [comment]
@mbernasocchi Just add in the metadata.txt, I think before @timlinux went to leave, he told me that he will manage the visual changelog

**Pull request merged**